virttest/vt_cluster/proxy.py

```diff
@@ -0,0 +1,166 @@
+                exc_type_name = fault_string.get("exc_type").split(".")[-1]
+            else:
+                root_mod_name = None
+                exc_type_name = fault_string.get("exc_type")
+
+            kargs = fault_string.get("exc_value")
+
+            # FIXME: For backward compatibility with the current framework interfaces,
+            #        this code reconstructs the remote exception locally. A better
+            #        approach would be to use a structured error code mechanism
+            #        instead of dynamically importing and reconstructing exceptions.
+            try:
+                if root_mod_name:
+                    actual_root_mod = importlib.import_module(root_mod_name)
+                    specific_exception_class = getattr(actual_root_mod, exc_type_name)
+                else:
+                    specific_exception_class = getattr(
+                        importlib.import_module("builtins"), exc_type_name, None
+                    )
+                    if specific_exception_class is None:
+                        specific_exception_class = eval(exc_type_name)
+
+                if isinstance(kargs, dict):
+                    raise specific_exception_class(**kargs)
+                elif isinstance(kargs, str):
+                    raise specific_exception_class(kargs)
+                else:
+                    raise specific_exception_class()
+            except Exception:
+                raise ServerProxyError(fault_code, fault_string.get("tb_info")) from e
+
+
+class _ClientProxy(client.ServerProxy):
+    """
+    An XML-RPC client proxy for communicating with a remote agent server.
+
+    It extends `xmlrpc.client.ServerProxy` to use `_ClientMethod` for
+    method calls, enabling more detailed logging and custom error handling,
+    including attempting to reconstruct remote exceptions.
+
+    :param uri: The URI of the remote XML-RPC server (agent server).
+    :type uri: str
+    """
+
+    def __init__(self, uri):
+        super(_ClientProxy, self).__init__(uri, allow_none=True, use_builtin_types=True)
+        self._uri = uri
+
+    def __getattr__(self, name):
+        return _ClientMethod(self._ServerProxy__request, name)
+
+    def __getstate__(self):
+        """
+        Custom pickle serialization to avoid connection attempts.
+
+        Only serialize the URI, not the transport or other connection objects.
+        """
+        return {"uri": self._uri}
+
+    def __setstate__(self, state):
+        """
+        Custom pickle de-serialized to reconstruct the proxy.
+
+        Reinitialize the proxy with the stored URI.
+        """
+        self.__init__(state["uri"])
+
+
+def get_server_proxy(uri):
+    """
+    Get the server proxy.
+
+    :param uri: The URI of the server proxy. e.g: http://$host:$proxy_port/
+    :type uri: str
+    :return: The proxy obj.
+    :rtype: _ClientProxy or _LocalProxy
```

Comment:
It seems no _LocalProxy will be returned

Reply:
Thanks, have updated!